vendor/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jsonmergepatch
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/evanphx/json-patch"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+)
+
+// Create a 3-way merge patch based-on JSON merge patch.
+// Calculate addition-and-change patch between current and modified.
+// Calculate deletion patch between original and modified.
+func CreateThreeWayJSONMergePatch(original, modified, current []byte, fns ...mergepatch.PreconditionFunc) ([]byte, error) {
+	if len(original) == 0 {
+		original = []byte(`{}`)
+	}
+	if len(modified) == 0 {
+		modified = []byte(`{}`)
+	}
+	if len(current) == 0 {
+		current = []byte(`{}`)
+	}
+
+	addAndChangePatch, err := jsonpatch.CreateMergePatch(current, modified)
+	if err != nil {
+		return nil, err
+	}
+	// Only keep addition and changes
+	addAndChangePatch, addAndChangePatchObj, err := keepOrDeleteNullInJsonPatch(addAndChangePatch, false)
+	if err != nil {
+		return nil, err
+	}
+
+	deletePatch, err := jsonpatch.CreateMergePatch(original, modified)
+	if err != nil {
+		return nil, err
+	}
+	// Only keep deletion
+	deletePatch, deletePatchObj, err := keepOrDeleteNullInJsonPatch(deletePatch, true)
+	if err != nil {
+		return nil, err
+	}
+
+	hasConflicts, err := mergepatch.HasConflicts(addAndChangePatchObj, deletePatchObj)
+	if err != nil {
+		return nil, err
+	}
+	if hasConflicts {
+		return nil, mergepatch.NewErrConflict(mergepatch.ToYAMLOrError(addAndChangePatchObj), mergepatch.ToYAMLOrError(deletePatchObj))
+	}
```

Comment:
Also, in general there are a lot of functions which return errors. I don't have a full grasp of the ways this is likely to fail in practice. It might be nice to have some tests that exercise some of these code paths.

Reply:
This code path is only taken for `ConfigMap`s, and only because we have chosen not to deal with them the say way `kubectl apply` would.

(It's a vendored library, too, so I don't think we'd want to change this file directly if we deem it a problem)

Reply:
> When it's not possible to patch an object without conflicts, what should the operator do? One concern I have is that we'll get in a state where we can never update an object and it basically goes unmaintained.

The error will bubble back up to the reconciler. Is there something special way you think it should be handled other than rate-limited retries?

Reply:
Hah, I had initially thought this was part of manifestival, so my previous comments are maybe not that relevant.

Rate-limited retries are great for transient failures, but I'm inclined to think that merge conflicts don't fall under that category. In the case of `kubectl apply`, it generally reports them back to you and expects you to figure out how to resolve them, which isn't great for an automated operator. It appears that `kubectl apply` has a `--overwrite=true` option which "Automatically resolve conflicts between the modified and live configuration by using values from
the modified configuration". That approach seems like a good fit for the operator to me, but I have no idea how it's implemented.

Thoughts?

Reply:
We could easily implement that behavior, I think: we'd just fall back to [creating a JSON merge patch](https://github.com/jcrossley3/manifestival/blob/2e2aa3de9981ae0363e40d1da46c92aaf4fe8828/patch/patch.go#L55) if the [strategic merge patch](https://github.com/jcrossley3/manifestival/blob/2e2aa3de9981ae0363e40d1da46c92aaf4fe8828/patch/patch.go#L64) resulted in a conflict.

Reply:
Sounds good. Will you write up an issue to track that and/or collect any feedback?

Reply:
Sure thing!